### PR TITLE
improve xor-linked-list in cpp and add list test

### PR DESCRIPTION
**Pull request by iattempt**
**Fixes issue:**

**Changes:**
in xor_linked_list.cpp:
1. implement 5 insertions according to std (passed test)
2. add erase function prototype (untested)
3. use another way to store sentinels to avoid lot of conditions in insertion or other functions implements
4. fix order of initialize error
5. fix coding style

in list_test.cpp:
1. rename and fix coding style
2. add more test for insertions
3. add type alias to be easier to  replace container which test to

### CODE REVIEW

File: code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp [hunk 1]
```diff
@@ -648,177 +740,3 @@ XorLinkedList<_Type, _Compare>::crend() const->const_reverse_iterator
...
-
-// Base base() const
-// {
-// return base_;
-// }
-
-// private:
-// Base base_;
-// friend XorLinkedList<value_type>;
-// };
```
Comment (iattempt) [inline code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: remove unused class

File: code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp [hunk 2]
```diff
@@ -25,24 +32,12 @@ class __Node
...
-    }
-
     inline value_type &value()
     {
         return value_;
     }
 
-    inline value_type const &value() const
+    inline const value_type &value() const
     {
```
Comment (arnavb) [inline code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: While I understand that you are using STL naming conventions, I still suggest you use camel case.

File: code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp [hunk 3]
```diff
@@ -55,13 +50,23 @@ class __Node
...
+        return reinterpret_cast<SPNode>(reinterpret_cast<uintptr_t>(around) ^
+                                        reinterpret_cast<uintptr_t>(around_));
+    }
 
-template<class _Type>
-class ListConstIter;
+    inline void around(const SPNode &prev, const SPNode &next)
+    {
+        this->around_ = reinterpret_cast<void *>(reinterpret_cast<uintptr_t>(prev) ^
+                                                 reinterpret_cast<uintptr_t>(next));
```
Comment (arnavb) [inline code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: Is the this necessary

Comment (iattempt) [reply code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: I'll remove it. 😃

File: code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp (hunk 2 above)
Comment (iattempt) [reply code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: Maybe some of alias is need to change to CamelCase, but the STL naming conventions is useful for users, I found this <a href="https://stackoverflow.com/questions/44571362/what-is-the-use-of-value-type-in-stl-containers">example</a> to demonstrate what I mean for.

Comment (arnavb) [reply code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: :+1:

File: code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp [hunk 4]
```diff
@@ -128,18 +133,19 @@ class ListIter :public std::iterator<std::bidirectional_iterator_tag, _Type>
...
+    bool operator==(const Self &other) const
     {
         return curr_ == other.curr_;
     }
 
-    bool operator!=(Self const &other) const
+    bool operator!=(const Self &other) const
     {
         return curr_ != other.curr_;
     }
```
Comment (arnavb) [inline code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: Better: return !(*this == other);

File: code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp [hunk 5]
```diff
@@ -210,18 +216,24 @@ class ListConstIter :public std::iterator<std::bidirectional_iterator_tag, _Type
...
 
-    bool operator==(Self const &other) const
+    bool operator==(const Self &other) const
     {
         return curr_ == other.curr_;
     }
 
-    bool operator!=(Self const &other) const
+    bool operator!=(const Self &other) const
     {
```
Comment (arnavb) [inline code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: Same here

File: code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp [hunk 6]
```diff
@@ -237,29 +249,33 @@ class XorLinkedList
     using value_type = _Type;
     using size_type = size_t;
     using difference_type = ptrdiff_t;
-    using pointer = value_type const *;
-    using const_pointer = value_type const *;
+    using pointer = const value_type *;
```
Comment (arnavb) [inline code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: Don’t put spaces between the asterisk/ampersand and the name

File: code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp [hunk 7]
```diff
@@ -354,7 +379,7 @@ XorLinkedList<_Type, _Compare>::front()->reference
     if (empty())
```
Comment (arnavb) [inline code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: How about using an exception class on the line below?

File: code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp [hunk 8]
```diff
@@ -422,56 +441,140 @@ XorLinkedList<_Type, _Compare>::pop_back()
...
 
 template<typename _Type, typename _Compare>
-inline void
-XorLinkedList<_Type, _Compare>::erase(iterator pos)
+inline auto
+XorLinkedList<_Type, _Compare>::insert(const_iterator pos, size_type size, const value_type &v)
+->iterator
 {
-    if (pos != end())
+    if (size == 0)
```
Comment (arnavb) [inline code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: if (empty())

Comment (arnavb) [reply code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: Discard this commen

File: code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp [hunk 9]
```diff
@@ -492,117 +595,108 @@ template<typename _Type, typename _Compare>
...
-    auto insteadNode = curr->around(prev);
-    auto nextOfInsteadNode = insteadNode->around(curr);
-    auto prevOfPrevNode = prev->around(curr);
-    insteadNode->around(prev, nextOfInsteadNode);
-    prev->around(prevOfPrevNode, insteadNode);
+    auto nextOfCurr = curr->around(prev);
+    auto nextOfNext = nextOfCurr->around(curr);
+    auto prevOfPrev = prev->around(curr);
+    nextOfCurr->around(prev, nextOfNext);
+    prev->around(prevOfPrev, nextOfCurr);
```
Comment (arnavb) [inline code/data_structures/linked_list/xor_linked_list/xor_linked_list.cpp]: A lot of places you have opportunities to use unique_ptr if you want.

File: code/data_structures/linked_list/list_test/xor_linked_list_test.cpp [hunk 10]
```diff
@@ -2142,6 +2136,42 @@ TEST_CASE("operations")
...
 }
 
+TEST_CASE("stl compatible")
+{
+    vectorContainer container = getRandomValueContainer();
+    expectListContainer expect;
+    actualListContainer actual;
+    expectListContainer::iterator expectPos;
+    actualListContainer::iterator actualPos;
+    int randomValue;
```
Comment (arnavb) [inline code/data_structures/linked_list/list_test/xor_linked_list_test.cpp]: Why is this here?

Comment (iattempt) [reply code/data_structures/linked_list/list_test/xor_linked_list_test.cpp]: Is the randomValue? or this TEST_CASE?

Comment (iattempt) [reply code/data_structures/linked_list/list_test/xor_linked_list_test.cpp]: The test is incomplete, and the variable may use.

Comment (arnavb) [reply code/data_structures/linked_list/list_test/xor_linked_list_test.cpp]: Ok

**Comment by AdiChat** [comment]
Looks great 😍 

Hope you are enjoying your journey 🚆  with **OpenGenus**
